general/mm.hpp

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_MM
+#define MFEM_MM
+
+#include <list>
+#include <cstddef> // for size_t
+#include <unordered_map>
+
+using std::size_t;
+
+#include "occa.hpp" // for OccaMemory
+
+namespace mfem
+{
+
+// Implementation of MFEM's lightweight host/device memory manager (mm) designed
+// to work seamlessly with the okina device kernel interface.
+
+/// The memory manager singleton
+class mm
+{
+public:
+   struct alias;
+
+   // TODO: Change this to ptr
+   struct memory
+   {
+      const size_t bytes;
+      void *const h_ptr;
+      void *d_ptr;
+      OccaMemory o_ptr;
+      std::list<const alias *> aliases;
+      bool host;
+      bool padding[7];
+      memory(void* const h, const size_t b):
+         bytes(b), h_ptr(h), d_ptr(nullptr), aliases(), host(true) {}
+   };
+
+   struct alias
+   {
+      memory *const mem;
+      const long offset;
+   };
+
+   typedef std::unordered_map<const void*, memory> memory_map;
+   typedef std::unordered_map<const void*, const alias*> alias_map;
+
+   struct ledger
+   {
+      memory_map memories;
+      alias_map aliases;
+   };
+
+   /// Main malloc template function. Allocates n*size bytes and returns a
+   /// pointer to the allocated memory.
+   template<class T>
+   static inline T* malloc(const size_t n, const size_t size = sizeof(T))
+   { return static_cast<T*>(MM().Insert(::new T[n], n*size)); }
+
+   /// Frees the memory space pointed to by ptr, which must have been returned
+   /// by a previous call to mm::malloc.
+   template<class T>
+   static inline void free(void *ptr)
```

Comment:
Given their implementation, shouldn't `mm::malloc` and `mm::free` be called `mm::new` and `mm::delete`?

Reply:
I would have liked, but I think we cannot overload these operators.

Reply:
That's definitely something to try.

Reply:
I think what @tzanio means is that these are calling the constructors and destructors (as `new` and `delete` do), whereas `malloc` and `free` only provide the memory.

I don't like the design of these functions. Having `mm::malloc` call the constructor on an array of objects makes defining pooled memory managers using placement new very challenging, as the array versions of placement new/delete by default don't call the constructors or destructors.

See (10) here: https://en.cppreference.com/w/cpp/memory/new/operator_new